core/src/commonMain/kotlin/com/xebia/functional/xef/prompt/expressions/Expression.kt

```diff
@@ -0,0 +1,98 @@
+package com.xebia.functional.xef.prompt.expressions
+
+import com.xebia.functional.xef.auto.CoreAIScope
+import com.xebia.functional.xef.auto.PromptConfiguration
+import com.xebia.functional.xef.llm.ChatWithFunctions
+import com.xebia.functional.xef.llm.models.chat.Message
+import com.xebia.functional.xef.llm.models.chat.Role
+import com.xebia.functional.xef.prompt.experts.ExpertSystem
+import io.github.oshai.kotlinlogging.KLogger
+import io.github.oshai.kotlinlogging.KotlinLogging
+
+class Expression(
+  private val scope: CoreAIScope,
+  private val model: ChatWithFunctions,
+  val block: suspend Expression.() -> Unit
+) {
+
+  private val logger: KLogger = KotlinLogging.logger {}
+
+  private val messages: MutableList<Message> = mutableListOf()
+
+  private val generationKeys: MutableList<String> = mutableListOf()
+
+  suspend fun system(message: suspend () -> String) {
+    messages.add(Message(role = Role.SYSTEM, content = message(), name = Role.SYSTEM.name))
+  }
+
+  suspend fun user(message: suspend () -> String) {
+    messages.add(Message(role = Role.USER, content = message(), name = Role.USER.name))
+  }
+
+  suspend fun assistant(message: suspend () -> String) {
+    messages.add(Message(role = Role.ASSISTANT, content = message(), name = Role.ASSISTANT.name))
+  }
+
+  fun prompt(key: String): String {
+    generationKeys.add(key)
+    return "{{$key}}"
+  }
+
+  suspend fun run(
+    promptConfiguration: PromptConfiguration = PromptConfiguration.DEFAULTS
+  ): ExpressionResult {
+    block()
+    val instructionMessage =
+      Message(
+        role = Role.USER,
+        content =
+          ExpertSystem(
+              system = "You are an expert in replacing variables in templates",
+              query =
+                """
+          |I want to replace the following variables in the following template:
+          |<template>
+          |${messages.joinToString("\n") { it.content }}
```

Comment:
I think that we are trying to replace values only in the last message, it doesn't make sense to replace placeholders in the previous messages because they are messages in the conversation

Can we send only the last message here? I think that we can reduce the number of tokens here

Reply:
We can use generation at any point. Only generation triggers value replacement. They can also be in an assistant message, so we need to process all messages.
For example:
```kotlin
system { "You are an expert role player creator for CyberPunk 2077 and know how to use tools ${prompt("cyberpunk_tools")}" }
```

Reply:
To say some other way, multiple Messages form a single template, or at least that is what this implementation does.